swap/src/protocol/alice/swap.rs

```diff
@@ -279,27 +279,68 @@ where
+                        // We might not be able to publish the redeem transaction on the first try due to any number of reasons
+                        let backoff = backoff::ExponentialBackoffBuilder::new()
+                            .with_max_elapsed_time(None)
+                            .with_max_interval(Duration::from_secs(60))
+                            .build();
+
+                        match backoff::future::retry(backoff, || async {
+                            // If the cancel timelock is expired, we do not need to publish anymore
+                            // We cannot use a tokio::select! here because this is not cancellation safe
+                            if !matches!(
+                                state3.expired_timelocks(bitcoin_wallet).await?,
+                                ExpiredTimelocks::None { .. }
+                            ) {
+                                return Ok(None);
                             }
-                        },
-                        Err(error) => {
-                            tracing::error!("Failed to publish redeem transaction: {:#}", error);
-                            tx_lock_status
-                                .wait_until_confirmed_with(state3.cancel_timelock)
-                                .await?;
-
-                            AliceState::CancelTimelockExpired {
-                                monero_wallet_restore_blockheight,
-                                transfer_proof,
-                                state3,
+
+                            bitcoin_wallet
+                                .broadcast(tx.clone(), "redeem")
+                                .await
+                                .inspect_err(|e| {
+                                    tracing::warn!(
+                                        swap_id = %swap_id,
+                                        error = ?e,
+                                        "Failed to broadcast Bitcoin redeem transaction. We will retry."
+                                    )
+                                })
+                                .map(Some)
+                                .map_err(backoff::Error::transient)
+                        })
+                        .await
+                        {
+                            // We successfully published the redeem transaction
+                            // We wait until we see the transaction in the mempool before transitioning to the next state
+                            Ok(Some((_, subscription))) => match subscription.wait_until_seen().await {
+                                Ok(_) => AliceState::BtcRedeemTransactionPublished { state3 },
+                                Err(e) => {
+                                    bail!("Waiting for Bitcoin redeem transaction to be in mempool failed with {}! The redeem transaction was published, but it is not ensured that the transaction was included! You're screwed.", e)
```

Comment:
might be worth including the txid and tx hex here, and telling the user they can attempt to submit it manually via mempool.space api?

Reply:
Good idea

Reply:
Done